src/shapes/sphere.cpp

```diff
@@ -402,35 +434,70 @@ class Sphere final : public Shape<Float, Spectrum> {
                                                      uint32_t recursion_depth,
                                                      Mask active) const override {
         MI_MASK_ARGUMENT(active);
+        constexpr bool IsDiff = dr::is_diff_v<Float>;
 
         // Early exit when tracing isn't necessary
         if (!m_is_instance && recursion_depth > 0)
             return dr::zeros<SurfaceInteraction3f>();
 
-        // Recompute ray intersection to get differentiable t
-        Float t = pi.t;
-        if constexpr (dr::is_diff_v<Float>)
-            t = dr::replace_grad(t, ray_intersect_preliminary(ray, active).t);
-
-        // TODO handle RayFlags::FollowShape and RayFlags::DetachShape
-
         // Fields requirement dependencies
-        bool need_dn_duv = has_flag(ray_flags, RayFlags::dNSdUV) ||
-                           has_flag(ray_flags, RayFlags::dNGdUV);
-        bool need_dp_duv = has_flag(ray_flags, RayFlags::dPdUV) || need_dn_duv;
-        bool need_uv     = has_flag(ray_flags, RayFlags::UV) || need_dp_duv;
+        bool need_dn_duv  = has_flag(ray_flags, RayFlags::dNSdUV) ||
+                            has_flag(ray_flags, RayFlags::dNGdUV);
+        bool need_dp_duv  = has_flag(ray_flags, RayFlags::dPdUV) || need_dn_duv;
+        bool need_uv      = has_flag(ray_flags, RayFlags::UV) || need_dp_duv;
+        bool detach_shape = has_flag(ray_flags, RayFlags::DetachShape);
+        bool follow_shape = has_flag(ray_flags, RayFlags::FollowShape);
+
+        const Point3f& center = m_center.value();
+        const Float& radius = m_radius.value();
+        const Transform4f& to_world = m_to_world.value();
+        const Transform4f& to_object = m_to_object.value();
+
+        /* If necessary, temporally suspend gradient tracking for all shape
+           parameters to construct a surface interaction completely detach from
+           the shape. */
+        dr::suspend_grad<Float> scope(detach_shape, center, radius, to_world, to_object);
```

Comment:
Throws error when trying to add a test for `DetachShape`

Reply:
This is fixed now in Dr.Jit